src/platform_impl/linux/x11/ime/context.rs

```diff
@@ -355,25 +354,44 @@ impl ImeContext {
         self.allowed
     }
 
-    // Set the spot for preedit text. Setting spot isn't working with libX11 when preedit callbacks
-    // are being used. Certain IMEs do show selection window, but it's placed in bottom left of the
-    // window and couldn't be changed.
-    //
-    // For me see: https://bugs.freedesktop.org/show_bug.cgi?id=1580.
-    pub(crate) fn set_spot(&mut self, xconn: &Arc<XConnection>, x: c_short, y: c_short) {
-        if !self.is_allowed() || self.ic_spot.x == x && self.ic_spot.y == y {
+    /// Set the spot and area for preedit text.
+    ///
+    /// This functionality depends on the libx11 version.
+    ///  - Until libx11 1.8.2, XNSpotLocation was blocked by libx11 in On-The-Spot mode.
+    ///  - Until libx11 1.8.11, XNArea was blocked by libx11 in On-The-Spot mode.
+    ///
+    /// Use of this information is discretionary by input method servers,
+    /// and some may not use it by default, even if they have support.
+    pub(crate) fn set_area(
+        &mut self,
+        xconn: &Arc<XConnection>,
+        x: i16,
+        y: i16,
+        width: u16,
+        height: u16,
+    ) {
+        if !self.is_allowed()
+            || self.ic_area.x == x
+                && self.ic_area.y == y
+                && self.ic_area.width == width
+                && self.ic_area.height == height
+        {
             return;
         }
 
-        self.ic_spot = ffi::XPoint { x, y };
+        self.ic_area = ffi::XRectangle { x, y, width, height };
+        let ic_spot =
+            ffi::XPoint { x: x.saturating_add(width as i16), y: y.saturating_add(height as i16) };
```

Comment:
Do we really need to `sub` like that? The point was that area would prevent that issue? Since you tell the area starting at top-left corner with w/h to not obscure.

Reply:
The area is correct, but the spot still needs to be somewhere, and for remaining pre-libx11-1.8.11, pre-fcitx5-5.1.11, or iBus until they merge the change system will go back to the previous (bad) behavior. Alacritty has a workaround for this reason (as seen in my patch in my branch), and would need to retain this workaround until XNArea-compatible XIM servers are the norm. Furthermore, this is how it is done on Windows as well (they have both an area and a spot, and the windows equivalent of the spot location is calculated the same way), and in the future it might be useful for right-to-left aware IM servers to use both the spot and the area.

Reply:
I'm marking this as resolved, but I'm happy to hear any ongoing issue with this, if I have missed your point.